psamm/randomsparse.py

```diff
@@ -0,0 +1,207 @@
+# along with PSAMM.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Copyright 2014-2015  Jon Lund Steffensen <jon_steffensen@uri.edu>
+# Copyright 2015  Keith Dufault-Thompson <keitht547@my.uri.edu>
+# Copyright 2016  Chao Liu <lcddzyx@gmail.com>
+
+from __future__ import unicode_literals
+
+import random
+import logging
+
+from . import fluxanalysis
+from .expression import boolean
+
+from six import iteritems, string_types
+
+logger = logging.getLogger(__name__)
+
+
+class ReactionDeletionStrategy(object):
+    """Deleting reactions strategy class.
+    """
+
+    def __init__(self, model, reaction_set=None):
+        self._model = model
+        if reaction_set is None:
+            self._test_set = set(self._model.reactions)
+        else:
+            self._test_set = set(reaction_set)
+        self._reactions = set(self._test_set)
+
+    def get_total(self):
+        return self._reactions
+
+    def tests(self):
+        while len(self._test_set) > 0:
+            reaction = random.choice(tuple(self._test_set))
+            self._test_set.remove(reaction)
+            yield reaction, {reaction}
+
+    def delete(self, entity, deleted_reactions):
+        pass
+
+
+class GeneDeletionStrategy(object):
+    """Deleting genes strategy class.
+    """
+
+    def __init__(self, model, gene_assoc):
+        self._model = model
+        self._genes = set()
+        self._gene_assoc = dict(gene_assoc)
+
+        for reaction, assoc in iteritems(self._gene_assoc):
+            self._genes.update(v.symbol for v in assoc.variables)
+
+        self._reactions = set(self._model.reactions)
+        self._test_set = set(self._genes)
+
+    def get_total(self):
+        return set(self._genes)
+
+    def tests(self):
+        while len(self._test_set) > 0:
+            gene = random.choice(tuple(self._test_set))
+            self._test_set.remove(gene)
+
+            deleted_reactions = set()
+            self._new_gene_assoc = {}
+            for reaction in self._reactions:
+                if reaction not in self._gene_assoc:
+                    continue
+                assoc = self._gene_assoc[reaction]
+                if boolean.Variable(gene) in assoc.variables:
+                    new_assoc = assoc.substitute(
+                        lambda v: v if v.symbol != gene else False)
+                    if new_assoc.has_value() and not new_assoc.value:
+                        deleted_reactions.add(reaction)
+                    else:
+                        self._new_gene_assoc[reaction] = new_assoc
+                else:
+                    self._new_gene_assoc[reaction] = assoc
+
+            yield gene, deleted_reactions
+
+    def delete(self, entity, deleted_reactions):
+        self._reactions -= deleted_reactions
+        self._gene_assoc = self._new_gene_assoc
+
+
+def get_gene_associations(model):
+    """Create gene association for class :class:`.GeneDeletionStrategy`.
+
+    Args:
```

Comment:
Add a line describing what it does, e.g. "Return a dict mapping reaction IDs to `boolean.Expression` objects representing the genes that the reaction depends on.".